docs/gene_sets.md

```diff
@@ -0,0 +1,136 @@
+
+* `gene_set_name`
+* `gene_set_description`
+* `gene_symbol`
+* `gene_description`
+
+Publishers MAY include additional header columns. It is RECOMMENDED that these custom columns observe the same self-documenting style - `column_name` and `column_description` for easier comprehension by data consumers.
+
+## Rows
+
+Each subsequent row describes a gene in a gene set.
+<br><br>
+
+### Values
+
+The values for `gene_set_name`, `gene_set_description`, and `gene_symbol` MUST NOT contain illegal ASCII characters or sequences. If the following cases are detected, validation MUST display a warning and fail the upload:
+
+* control characters (decimal 0-31)
+* DEL (decimal 127)
+* leading spaces (decimal 32) in a field - "     This is an example"
+* trailing spaces (decimal 32) in a field - "This is an example     " 
+* multiple spaces (decimal 32) "internal" to a field - "This     is an example"
+
+### `gene_set_name`
+
+---
+
+The `gene_set_name` column MUST contain a value.
+
+If the `gene_set_name` is missing, validation MUST display a warning and fail the upload. This is illustrated by **~~?~~** in the example:
+
+| gene_set_name | gene_set_description | gene_symbol | gene_description |
+|---------------|----------------------|-------------|------------------|
+| club.cell     | description          | CCKAR       | description      |
+|     **~~?~~** |                      | SCGB3A2     | description      |
+| club.cell     |                      | CYP2F2      | description      |
+
+
+All `gene_symbol(s)` for a `gene_set_name` MUST be on contiguous rows; otherwise, an out-of-order row is interpreted as a duplicate `gene_set_name`. In this case, validation MUST display a warning and fail the upload. This is illustrated by **~~club.cell~~** in the example:
+
+---
+| gene_set_name | gene_set_description | gene_symbol | gene_description |
+|---------------|----------------------|-------------|------------------|
+| club.cell     | description          | CCKAR       | description      |
+| club.cell     |                      | SCGB3A2     | description      |
+| club.cell     |                      | CYP2F2      | description      |
+| macrophage    | description          | CD68        | description      |
+| **~~club.cell~~**     |                      | CD163       | description      |
+
+
+When new gene sets are being added to a data collection on the portal, validation MUST detect `gene_set_name` collisions with current gene sets in the collection, display a warning, and fail the upload.  <br><br>
+
+### `gene_set_description`
+
+---
+
+The first instance of a `gene_set_description` column for a specific `gene_set_name` MUST contain a value. All other instances are ignored in subsequent rows for the same `gene_set_name`.
+
+If the first instance of the `gene_set_description` is missing, validation MUST display a warning and fail the upload. This is illustrated by **~~?~~** in the example:
```

Comment:
I believe gene set description should be optional.

Reply:
I'm unsure how to interpret the [desktop requirements:](https://docs.google.com/document/d/1O-Os0Bg6a9FBKAA1KIMR12KESAEhj2pD32lJFC5OGko/edit#)

_gene_set_description (**a field cannot be null**)
The **first non null field** for gene_set_description, all other fields will be ignored_

which reads like at least one `gene_set_description` for a gene set MUST NOT be null which is not quite optional.

Reply:
Is the requirement:

At least one `gene_set_description` in a `gene_set` MUST have a valid (non-null) value; otherwise, validation fails.. The first valid field for `gene_set_description` sets the gene set description for the gene set. Further `gene_set_description` fields for the `gene_set` on subsequent rows MUST be ignored.

Reply:
the quoted prose is obsolete (my mistake) - I believe that in discussions with Signe we changed it to better support creator (not just final publication) scenarios.   That said, you guys are the decision maker here, and I think we could go either way.

Reply:
Please look at the REST API spec - it is more accurate, and reflects the final decisions Signe made.  She did not have a chance to fully update the desktop requirements doc.

Reply:
The [REST API](https://docs.google.com/document/d/12Wce-r2OvLGnuUWm9abAiZuj-kELz_LYTEvv6KZLXXo/edit#) states [my emphasis]:

> Where a gene set has multiple gene rows, and they differ in gene_set_description value, **any implementation may be used to select an appropriate gene set description.** **Desktop semantics defined** [here](https://docs.google.com/document/d/1O-Os0Bg6a9FBKAA1KIMR12KESAEhj2pD32lJFC5OGko/edit#) (see section 1F).

which returns me to the specification that I quoted above.

In this case, I guess _Portal semantics are what I wrote in this document._

Reply:
@brianraymor your interpretation and rewording of what I wrote in the desktop requirements is correct (and is worded much more clearly). I agree with you that requiring gene_set_description for the portal is best for the users since the end user of gene sets will likely be different than the people who created them and the description will provide more context.

@bkmartinjr for desktop, I agree that gene_set_description should be optional since the users creating the gene sets are more likely to also be the ones exploring them in the explorer. I'll update the requirements doc to reflect this (and point out the difference with portal).

For posterity, I don't see the difference in requirements between desktop and portal for gene_set_description as problematic. A user may want to download gene sets from explorer and publish these gene sets in portal but at that point adding a description is in their best interest. This workflow happens infrequently (publishing data doesn't happen very often) and there are multiple steps a user needs to take before publishing data anyway.